usaspending_api/common/etl/spark.py

```diff
@@ -560,3 +564,179 @@ def create_ref_temp_views(spark: SparkSession, create_broker_views: bool = False
+            a merged file that was generated during the copy merge.
+    """
+    overwrite = True
+    if not logger:
+        logger = get_jvm_logger(spark)
+    hadoop = spark.sparkContext._jvm.org.apache.hadoop
+    conf = spark.sparkContext._jsc.hadoopConfiguration()
+    parts_dir_path = hadoop.fs.Path(parts_dir)
+
+    # See comment in docstring about how FileSystem is determined
+    fs = parts_dir_path.getFileSystem(conf)
+
+    if not fs.exists(parts_dir_path):
+        raise ValueError("Source directory {} does not exist".format(parts_dir))
+
+    file = parts_dir
+    file_path = hadoop.fs.Path(file)
+
+    # Don't delete first if disallowing overwrite.
+    if not overwrite and fs.exists(file_path):
+        raise Py4JError(
+            spark._jvm.org.apache.hadoop.fs.FileAlreadyExistsException(f"{str(file_path)} " f"already exists")
+        )
+    part_files = []
+
+    for f in fs.listStatus(parts_dir_path):
+        if f.isFile():
+            file_path = f.getPath()
+            if file_path.getName().startswith("_"):
+                logger.debug(f"Skipping non-part file: {file_path.getName()}")
+                continue
+            logger.debug(f"Including part file: {file_path.getName()}")
+            part_files.append(f.getPath())
+    if not part_files:
+        logger.warn("Source directory is empty with no part files. Attempting creation of file with CSV header only")
+        out_stream = None
+        try:
+            out_stream = fs.create(hadoop.fs.Path(parts_dir), overwrite)
+            out_stream.writeBytes(header + "\n")
+        finally:
+            if out_stream is not None:
+                out_stream.close()
+        return
+
+    part_files.sort(key=lambda f: str(f))  # put parts in order by part number for merging
+    paths_to_merged_files = []
+    for parts_file_group in _merge_grouper(part_files, max_rows_per_merged_file):
+        part_suffix = f"_{str(parts_file_group.part).zfill(2)}" if parts_file_group.part else ""
+        partial_merged_file = f"{parts_dir}.partial{part_suffix}"
+        partial_merged_file_path = hadoop.fs.Path(partial_merged_file)
+        merged_file_name = parts_dir + part_suffix + f".{file_format}"
+        paths_to_merged_files.append(merged_file_name)
+        merged_file_path = hadoop.fs.Path(merged_file_name)
+        if overwrite and fs.exists(merged_file_path):
+            fs.delete(merged_file_path, True)
+        out_stream = None
+        try:
+            if fs.exists(partial_merged_file_path):
+                fs.delete(partial_merged_file_path, True)
+            out_stream = fs.create(partial_merged_file_path)
+            out_stream.writeBytes(header + "\n")
+            _merge_file_parts(fs, out_stream, conf, hadoop, partial_merged_file_path, parts_file_group.file_list)
+        finally:
+            if out_stream is not None:
+                out_stream.close()
+        try:
+            fs.rename(partial_merged_file_path, merged_file_path)
```

Comment:
This is from your code

Reply:
parts_dir is the path to the parts directory

Reply:
> Definitely would be good to do some operational testing for LARGE (many 100s of part files) data such that it would yield many *_1.csv ... *_2.csv ... combined files

This is not within the scope of this story

Reply:
This should be fine if you protect against `parts_dir` coming in with a trailing `/` e.g. by using `.rstrip("/")`. You can resolve after fixed.

Reply:
addressed.